include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,1244 @@
+
+        // evalute residual
+        return residual(x, f);
+      },
+      [&](const VectorType &x) -> int {
+        Assert(
+          setup_jacobian,
+          ExcMessage(
+            "No setup_jacobian function has been attached to the NOXSolver object."));
+
+        // setup Jacobian
+        int flag = setup_jacobian(x);
+
+        if (flag != 0)
+          return flag;
+
+        if (setup_preconditioner)
+          {
+            // check if preconditioner needs to be updated
+            bool update_preconditioner =
+              ((additional_data.threshold_nonlinear_iterations > 0) &&
+               ((n_nonlinear_iterations %
+                 additional_data.threshold_nonlinear_iterations) == 0)) ||
+              (solve_with_jacobian_and_track_n_linear_iterations &&
+               (n_last_linear_iterations >
+                additional_data.threshold_n_linear_iterations));
+
+            if ((update_preconditioner == false) &&
+                (update_preconditioner_predicate != nullptr))
+              update_preconditioner = update_preconditioner_predicate();
+
+            if (update_preconditioner) // update preconditioner
+              flag = setup_preconditioner(x);
+          }
+
+        return flag;
+      },
+      [&](const VectorType &x, VectorType &v) -> int {
+        Assert(
+          apply_jacobian,
+          ExcMessage(
+            "No apply_jacobian function has been attached to the NOXSolver object."));
+
+        n_jacobian_applications++;
+
+        // apply Jacobian
+        return apply_jacobian(x, v);
+      },
+      [&](const VectorType &f, VectorType &x, const double tolerance) -> int {
+        n_nonlinear_iterations++;
+
+        // invert Jacobian
+        if (solve_with_jacobian)
+          {
+            // without tracking of linear iterations
+            return solve_with_jacobian(f, x, tolerance);
+          }
+        else if (solve_with_jacobian_and_track_n_linear_iterations)
+          {
+            // with tracking of linear iterations
+            const int n_linear_iterations =
+              solve_with_jacobian_and_track_n_linear_iterations(f,
+                                                                x,
+                                                                tolerance);
+
+            if (n_linear_iterations == -1)
+              return 1;
+
+            this->n_last_linear_iterations = n_linear_iterations;
+
+            return 0;
+          }
+        else
+          {
+            Assert(
+              false,
+              ExcMessage(
+                "Neither a solve_with_jacobian or a "
+                "solve_with_jacobian_and_track_n_linear_iterations function "
+                "has been attached to the NOXSolver object."));
+
+            Assert(false, ExcNotImplemented());
+            return 1;
+          }
+      }));
+
+    // setup solver control
+    auto check =
+      Teuchos::rcp(new NOX::StatusTest::Combo(NOX::StatusTest::Combo::OR));
```

Comment:
Probably, it would make sense to have a possibility to provide a completely user-defined check function. But I assume this could be done in a subsequent improvement of the wrapper.